src/js/game/LevelMVC/LevelModel.js

```diff
@@ -803,31 +803,33 @@ module.exports = class LevelModel {
     this.moveForward();
   }
 
-  placeBlock(blockType) {
-    const position = this.player.position;
-    let shouldPlace = false;
+  placeBlock(blockType, entity = this.player) {
+    const position = entity.position;
     let placedBlock = null;
 
-    switch (blockType) {
-      case "cropWheat":
-        shouldPlace = this.groundPlane.getBlockAt(position).blockType === "farmlandWet";
-        break;
-
-      default:
-        shouldPlace = true;
-        break;
-    }
-
-    if (shouldPlace === true) {
+    let ground = this.groundPlane.getBlockAt(position);
+    let result = entity.canPlaceBlockOver(blockType, ground.blockType);
+    if (result.canPlace) {
       var block = new LevelBlock(blockType);
-
-      placedBlock = this.actionPlane.setBlockAt(position, block);
-      this.player.isOnBlock = !block.isWalkable;
+      switch (result.plane) {
+        case "actionPlane":
+          placedBlock = this.actionPlane.setBlockAt(position, block);
+          entity.walkableCheck(block);
+          break;
+        case "groundPlane":
+          this.groundPlane.setBlockAt(position, block);
+          this.controller.levelView.refreshGroundPlane();
```

Comment:
why is it only necessary to refresh the ground plane, not the action plane?

Reply:
because the actionPlane isn't changing?

Reply:
I mean in the `"actionPlane"` case

Reply:
Oh wait, you totally mean why in this function. It's because when the actionPlane is changed, we don't refresh the whole thing. We only refresh a few local indices and that's done in setBlockAt()

We could totally add a general refreshGroundPlane to setBlokAt(). I'll do it that way :)